src/lem/eval.rs

```diff
@@ -446,14 +452,50 @@ fn apply_cont() -> Func {
             }
         }
     });
+    // Returns 2 if both arguments are U64, 1 if both arguments are numbers, and 0 otherwise
+    let args_num_type = func!(args_num_type(arg1, arg2): 1 => {
+        let other = Num(0);
+        match arg1.tag {
+            Expr::Num => {
+                match arg2.tag {
+                    Expr::Num => {
+                        let ret = Num(1);
+                        return (ret)
+                    }
+                    Expr::U64 => {
+                        let ret = Num(1);
+                        return (ret)
+                    }
+                };
+                return (other)
+            }
+            Expr::U64 => {
+                match arg2.tag {
+                    Expr::Num => {
+                        let ret = Num(1);
+                        return (ret)
+                    }
+                    Expr::U64 => {
+                        let ret = Num(2);
+                        return (ret)
+                    }
+                };
+                return (other)
+            }
+        };
+        return (other)
+    });
     func!(apply_cont(result, env, cont, ctrl): 4 => {
         // Useful constants
         let ret: Ctrl::Return;
         let makethunk: Ctrl::MakeThunk;
         let errctrl: Ctrl::Error;
         let err: Cont::Error;
-        let nil: Expr::Nil;
+        let nil = Symbol("nil");
+        let nil = cast(nil, Expr::Nil);
         let t = Symbol("t");
+        let zero = Num(0);
+        let size_u64 = Num(18446744073709551616);
```

Comment:
Eventually, yes. As of now, it will add constraints. I'm working on another set of gadgets which will allow constants to not be allocated: https://github.com/lurk-lab/lurk-rs/tree/lem-gadgets

The main datatype is `LurkNum` which is either a constant or a linear combination. The reason for adding the constant and not having just linear combinations (which can subsume constants), is because with constants you know you don't need to allocate when multiplying